Revert WebKitWebAudioSourceGStreamer changes that require GStreamer 1.16,
as that's too new for us.

Author: Emilio Pozuelo Monfort <pochu@debian.org>

diff --git a/Source/WebCore/platform/audio/gstreamer/WebKitWebAudioSourceGStreamer.cpp b/Source/WebCore/platform/audio/gstreamer/WebKitWebAudioSourceGStreamer.cpp
index c6e5bf0a9a..0b4bccfd5d 100644
--- a/Source/WebCore/platform/audio/gstreamer/WebKitWebAudioSourceGStreamer.cpp
+++ b/Source/WebCore/platform/audio/gstreamer/WebKitWebAudioSourceGStreamer.cpp
@@ -64,11 +64,13 @@ struct _WebKitWebAudioSrcPrivate {
     guint framesToPull;
     guint bufferSize;
 
+    GRefPtr<GstElement> interleave;
+
     GRefPtr<GstTask> task;
     GRecMutex mutex;
 
-    GRefPtr<GstElement> source;
-    GRefPtr<GstCaps> caps;
+    // List of appsrc. One appsrc for each planar audio channel.
+    Vector<GRefPtr<GstElement>> sources;
 
     // src pad of the element, interleaved wav data is pushed to it.
     GstPad* sourcePad;
@@ -115,6 +117,14 @@ static void webKitWebAudioSrcGetProperty(GObject*, guint propertyId, GValue*, GP
 static GstStateChangeReturn webKitWebAudioSrcChangeState(GstElement*, GstStateChange);
 static void webKitWebAudioSrcRenderIteration(WebKitWebAudioSrc*);
 
+static GstCaps* getGStreamerMonoAudioCaps(float sampleRate)
+{
+    return gst_caps_new_simple("audio/x-raw", "rate", G_TYPE_INT, static_cast<int>(sampleRate),
+        "channels", G_TYPE_INT, 1,
+        "format", G_TYPE_STRING, GST_AUDIO_NE(F32),
+        "layout", G_TYPE_STRING, "interleaved", nullptr);
+}
+
 static GstAudioChannelPosition webKitWebAudioGStreamerChannelPosition(int channelIndex)
 {
     GstAudioChannelPosition position = GST_AUDIO_CHANNEL_POSITION_NONE;
@@ -145,27 +155,6 @@ static GstAudioChannelPosition webKitWebAudioGStreamerChannelPosition(int channe
     return position;
 }
 
-static GstCaps* getGStreamerAudioCaps(float sampleRate, unsigned numberOfChannels)
-{
-    guint64 channelMask = 0;
-    Vector<GstAudioChannelPosition> positions;
-    positions.reserveInitialCapacity(numberOfChannels);
-
-    for (unsigned channelIndex = 0; channelIndex < numberOfChannels; channelIndex++) {
-        GstAudioChannelPosition position = webKitWebAudioGStreamerChannelPosition(channelIndex);
-        positions.uncheckedAppend(WTFMove(position));
-    }
-
-    gst_audio_channel_positions_to_mask(reinterpret_cast<GstAudioChannelPosition*>(positions.data()),
-        numberOfChannels, FALSE, &channelMask);
-
-    return gst_caps_new_simple("audio/x-raw", "rate", G_TYPE_INT, static_cast<int>(sampleRate),
-        "channels", G_TYPE_INT, static_cast<int>(numberOfChannels),
-        "channel-mask", GST_TYPE_BITMASK, channelMask,
-        "format", G_TYPE_STRING, GST_AUDIO_NE(F32),
-        "layout", G_TYPE_STRING, "non-interleaved", nullptr);
-}
-
 #define webkit_web_audio_src_parent_class parent_class
 WEBKIT_DEFINE_TYPE_WITH_CODE(WebKitWebAudioSrc, webkit_web_audio_src, GST_TYPE_BIN, GST_DEBUG_CATEGORY_INIT(webkit_web_audio_src_debug, "webkitwebaudiosrc", 0, "webaudiosrc element"))
 
@@ -186,24 +175,22 @@ static void webkit_web_audio_src_class_init(WebKitWebAudioSrcClass* webKitWebAud
     GParamFlags flags = static_cast<GParamFlags>(G_PARAM_CONSTRUCT_ONLY | G_PARAM_READWRITE);
     g_object_class_install_property(objectClass,
                                     PROP_RATE,
-                                    g_param_spec_float("rate",
-                                                       nullptr, nullptr,
-                                                       G_MINDOUBLE, G_MAXDOUBLE,
+                                    g_param_spec_float("rate", "rate",
+                                                       "Sample rate", G_MINDOUBLE, G_MAXDOUBLE,
                                                        44100.0, flags));
 
     g_object_class_install_property(objectClass,
                                     PROP_BUS,
-                                    g_param_spec_pointer("bus",
-                                                         nullptr, nullptr,
-                                                         flags));
+                                    g_param_spec_pointer("bus", "bus",
+                                                         "Bus", flags));
 
     g_object_class_install_property(objectClass, PROP_DESTINATION, g_param_spec_pointer("destination", "destination",
         "Destination", flags));
 
     g_object_class_install_property(objectClass,
                                     PROP_FRAMES,
-                                    g_param_spec_uint("frames",
-                                                      nullptr, nullptr,
+                                    g_param_spec_uint("frames", "frames",
+                                                      "Number of audio frames to pull at each iteration",
                                                       0, G_MAXUINT8, AudioUtilities::renderQuantumSize, flags));
 }
 
@@ -226,17 +213,40 @@ static void webKitWebAudioSrcConstructed(GObject* object)
     priv->task = adoptGRef(gst_task_new(reinterpret_cast<GstTaskFunction>(webKitWebAudioSrcRenderIteration), src, nullptr));
     gst_task_set_lock(priv->task.get(), &priv->mutex);
 
-    priv->source = makeGStreamerElement("appsrc", "webaudioSrc");
-    priv->caps = adoptGRef(getGStreamerAudioCaps(priv->sampleRate, priv->bus->numberOfChannels()));
+    priv->interleave = makeGStreamerElement("audiointerleave", nullptr);
+
+    if (!priv->interleave) {
+        GST_ERROR_OBJECT(src, "Failed to create audiointerleave");
+        return;
+    }
+
+    gst_bin_add(GST_BIN(src), priv->interleave.get());
+
+    // For each channel of the bus create a new upstream branch for interleave, like:
+    // appsrc ! . which is plugged to a new interleave request sinkpad.
+    for (unsigned channelIndex = 0; channelIndex < priv->bus->numberOfChannels(); channelIndex++) {
+        GUniquePtr<gchar> appsrcName(g_strdup_printf("webaudioSrc%u", channelIndex));
+        GRefPtr<GstElement> appsrc = makeGStreamerElement("appsrc", appsrcName.get());
+        GRefPtr<GstCaps> monoCaps = adoptGRef(getGStreamerMonoAudioCaps(priv->sampleRate));
+
+        GstAudioInfo info;
+        gst_audio_info_from_caps(&info, monoCaps.get());
+        GST_AUDIO_INFO_POSITION(&info, 0) = webKitWebAudioGStreamerChannelPosition(channelIndex);
+        GRefPtr<GstCaps> caps = adoptGRef(gst_audio_info_to_caps(&info));
 
-    // Configure the appsrc for minimal latency.
-    g_object_set(priv->source.get(), "max-bytes", static_cast<guint64>(2 * priv->bufferSize * priv->bus->numberOfChannels()), "block", TRUE,
-        "blocksize", priv->bufferSize,
-        "format", GST_FORMAT_TIME, "caps", priv->caps.get(), nullptr);
+        // Configure the appsrc for minimal latency.
+        g_object_set(appsrc.get(), "max-bytes", static_cast<guint64>(2 * priv->bufferSize), "block", TRUE,
+            "blocksize", priv->bufferSize,
+            "format", GST_FORMAT_TIME, "caps", caps.get(), nullptr);
 
-    gst_bin_add(GST_BIN(src), priv->source.get());
-    // appsrc's src pad is the only visible pad of our element.
-    GRefPtr<GstPad> targetPad = adoptGRef(gst_element_get_static_pad(priv->source.get(), "src"));
+        priv->sources.append(appsrc);
+
+        gst_bin_add(GST_BIN(src), appsrc.get());
+        gst_element_link_pads_full(appsrc.get(), "src", priv->interleave.get(), "sink_%u", GST_PAD_LINK_CHECK_NOTHING);
+    }
+
+    // interleave's src pad is the only visible pad of our element.
+    GRefPtr<GstPad> targetPad = adoptGRef(gst_element_get_static_pad(priv->interleave.get(), "src"));
     gst_ghost_pad_set_target(GST_GHOST_PAD(priv->sourcePad), targetPad.get());
 }
 
@@ -289,7 +299,7 @@ static void webKitWebAudioSrcGetProperty(GObject* object, guint propertyId, GVal
     }
 }
 
-static GRefPtr<GstBuffer> webKitWebAudioSrcAllocateBuffers(WebKitWebAudioSrc* src)
+static std::optional<Vector<GRefPtr<GstBuffer>>> webKitWebAudioSrcAllocateBuffers(WebKitWebAudioSrc* src)
 {
     WebKitWebAudioSrcPrivate* priv = src->priv;
 
@@ -298,37 +308,37 @@ static GRefPtr<GstBuffer> webKitWebAudioSrcAllocateBuffers(WebKitWebAudioSrc* sr
     if (!priv->destination || !priv->bus) {
         GST_ELEMENT_ERROR(src, CORE, FAILED, ("Internal WebAudioSrc error"), ("Can't start without destination or bus"));
         gst_task_stop(src->priv->task.get());
-        return nullptr;
+        return std::nullopt;
     }
 
     ASSERT(priv->pool);
 
-    GRefPtr<GstBuffer> buffer;
-    GstFlowReturn ret = gst_buffer_pool_acquire_buffer(priv->pool.get(), &buffer.outPtr(), nullptr);
-    if (ret != GST_FLOW_OK) {
-        // FLUSHING and EOS are not errors.
-        if (ret < GST_FLOW_EOS || ret == GST_FLOW_NOT_LINKED)
-            GST_ELEMENT_ERROR(src, CORE, PAD, ("Internal WebAudioSrc error"), ("Failed to allocate buffer for flow: %s", gst_flow_get_name(ret)));
-        return nullptr;
-    }
-
-    ASSERT(buffer);
-    ASSERT(priv->caps);
-
-    // Attach audio meta on buffer.
-    GstAudioInfo info;
-    gst_audio_info_from_caps(&info, priv->caps.get());
-    gst_buffer_add_audio_meta(buffer.get(), &info, priv->framesToPull, nullptr);
+    Vector<GRefPtr<GstBuffer>> channelBufferList;
+    channelBufferList.reserveInitialCapacity(priv->sources.size());
+    Vector<GstMappedBuffer> mappedBuffers;
+    mappedBuffers.reserveInitialCapacity(priv->sources.size());
+    for (unsigned i = 0; i < priv->sources.size(); ++i) {
+        GRefPtr<GstBuffer> buffer;
+        GstFlowReturn ret = gst_buffer_pool_acquire_buffer(priv->pool.get(), &buffer.outPtr(), nullptr);
+        if (ret != GST_FLOW_OK) {
+            // FLUSHING and EOS are not errors.
+            if (ret < GST_FLOW_EOS || ret == GST_FLOW_NOT_LINKED)
+                GST_ELEMENT_ERROR(src, CORE, PAD, ("Internal WebAudioSrc error"), ("Failed to allocate buffer for flow: %s", gst_flow_get_name(ret)));
+            return std::nullopt;
+        }
 
-    GstMappedBuffer mappedBuffer(buffer.get(), GST_MAP_READWRITE);
-    ASSERT(mappedBuffer);
-    for (unsigned channelIndex = 0; channelIndex < priv->bus->numberOfChannels(); channelIndex++)
-        priv->bus->setChannelMemory(channelIndex, reinterpret_cast<float*>(mappedBuffer.data() + channelIndex * priv->bufferSize), priv->framesToPull);
+        ASSERT(buffer);
+        GstMappedBuffer mappedBuffer(buffer.get(), GST_MAP_READWRITE);
+        ASSERT(mappedBuffer);
+        mappedBuffers.uncheckedAppend(WTFMove(mappedBuffer));
+        priv->bus->setChannelMemory(i, reinterpret_cast<float*>(mappedBuffers[i].data()), priv->framesToPull);
+        channelBufferList.uncheckedAppend(WTFMove(buffer));
+    }
 
-    return buffer;
+    return std::make_optional(channelBufferList);
 }
 
-static void webKitWebAudioSrcRenderAndPushFrames(const GRefPtr<GstElement>& element, GRefPtr<GstBuffer>&& buffer)
+static void webKitWebAudioSrcRenderAndPushFrames(GRefPtr<GstElement>&& element, Vector<GRefPtr<GstBuffer>>&& channelBufferList)
 {
     auto* src = WEBKIT_WEB_AUDIO_SRC(element.get());
     auto* priv = src->priv;
@@ -343,6 +353,8 @@ static void webKitWebAudioSrcRenderAndPushFrames(const GRefPtr<GstElement>& elem
     if (priv->hasRenderedAudibleFrame && !priv->destination->isPlaying())
         return;
 
+    ASSERT(channelBufferList.size() == priv->sources.size());
+
     GstClockTime timestamp = gst_util_uint64_scale(priv->numberOfSamples, GST_SECOND, priv->sampleRate);
     priv->numberOfSamples += priv->framesToPull;
     GstClockTime duration = gst_util_uint64_scale(priv->numberOfSamples, GST_SECOND, priv->sampleRate) - timestamp;
@@ -362,27 +374,36 @@ static void webKitWebAudioSrcRenderAndPushFrames(const GRefPtr<GstElement>& elem
         priv->hasRenderedAudibleFrame = true;
     }
 
-    GST_BUFFER_TIMESTAMP(buffer.get()) = outputTimestamp.position.nanoseconds();
-    GST_BUFFER_DURATION(buffer.get()) = duration;
-
-    if (priv->bus->isSilent())
-        GST_BUFFER_FLAG_SET(buffer.get(), GST_BUFFER_FLAG_GAP);
-
-    // Leak the buffer ref, because gst_app_src_push_buffer steals it.
-    GstFlowReturn ret = gst_app_src_push_buffer(GST_APP_SRC(priv->source.get()), buffer.leakRef());
-    if (ret != GST_FLOW_OK) {
-        // FLUSHING and EOS are not errors.
-        if (ret < GST_FLOW_EOS || ret == GST_FLOW_NOT_LINKED)
-            GST_ELEMENT_ERROR(src, CORE, PAD, ("Internal WebAudioSrc error"), ("Failed to push buffer on %s flow: %s", GST_OBJECT_NAME(priv->source.get()), gst_flow_get_name(ret)));
-        gst_task_stop(priv->task.get());
+    bool failed = false;
+    for (unsigned i = 0; i < priv->sources.size(); ++i) {
+        auto& buffer = channelBufferList[i];
+        GST_BUFFER_TIMESTAMP(buffer.get()) = outputTimestamp.position.nanoseconds();
+        GST_BUFFER_DURATION(buffer.get()) = duration;
+
+        if (priv->bus->channel(i)->isSilent())
+            GST_BUFFER_FLAG_SET(buffer.get(), GST_BUFFER_FLAG_GAP);
+
+        if (failed)
+            continue;
+
+        auto& appsrc = priv->sources[i];
+        // Leak the buffer ref, because gst_app_src_push_buffer steals it.
+        GstFlowReturn ret = gst_app_src_push_buffer(GST_APP_SRC(appsrc.get()), buffer.leakRef());
+        if (ret != GST_FLOW_OK) {
+            // FLUSHING and EOS are not errors.
+            if (ret < GST_FLOW_EOS || ret == GST_FLOW_NOT_LINKED)
+                GST_ELEMENT_ERROR(src, CORE, PAD, ("Internal WebAudioSrc error"), ("Failed to push buffer on %s flow: %s", GST_OBJECT_NAME(appsrc.get()), gst_flow_get_name(ret)));
+            gst_task_stop(priv->task.get());
+            failed = true;
+        }
     }
 }
 
 static void webKitWebAudioSrcRenderIteration(WebKitWebAudioSrc* src)
 {
     auto* priv = src->priv;
-    auto buffer = webKitWebAudioSrcAllocateBuffers(src);
-    if (!buffer) {
+    auto channelBufferList = webKitWebAudioSrcAllocateBuffers(src);
+    if (!channelBufferList) {
         gst_task_stop(priv->task.get());
         return;
     }
@@ -398,12 +419,10 @@ static void webKitWebAudioSrcRenderIteration(WebKitWebAudioSrc* src)
     Locker locker { AdoptLock, priv->dispatchToRenderThreadLock };
 
     if (!priv->dispatchToRenderThreadFunction)
-        webKitWebAudioSrcRenderAndPushFrames(GRefPtr<GstElement>(GST_ELEMENT_CAST(src)), WTFMove(buffer));
+        webKitWebAudioSrcRenderAndPushFrames(GRefPtr<GstElement>(GST_ELEMENT_CAST(src)), WTFMove(*channelBufferList));
     else {
-        priv->dispatchToRenderThreadFunction([buffer = WTFMove(buffer), protectedThis = GRefPtr<GstElement>(GST_ELEMENT_CAST(src))]() mutable {
-            if (!protectedThis)
-                return;
-            webKitWebAudioSrcRenderAndPushFrames(protectedThis, WTFMove(buffer));
+        priv->dispatchToRenderThreadFunction([channels = WTFMove(*channelBufferList), protectedThis = GRefPtr<GstElement>(GST_ELEMENT_CAST(src))]() mutable {
+            webKitWebAudioSrcRenderAndPushFrames(WTFMove(protectedThis), WTFMove(channels));
         });
     }
 
@@ -424,19 +443,13 @@ static GstStateChangeReturn webKitWebAudioSrcChangeState(GstElement* element, Gs
 
     switch (transition) {
     case GST_STATE_CHANGE_NULL_TO_READY:
-        priv->numberOfSamples = 0;
-        break;
-    case GST_STATE_CHANGE_READY_TO_PAUSED: {
-        priv->pool = adoptGRef(gst_buffer_pool_new());
-        GstStructure* config = gst_buffer_pool_get_config(priv->pool.get());
-        gst_buffer_pool_config_set_params(config, nullptr, priv->bufferSize * priv->bus->numberOfChannels(), 0, 0);
-        gst_buffer_pool_set_config(priv->pool.get(), config);
-        if (!gst_buffer_pool_set_active(priv->pool.get(), TRUE))
-            return GST_STATE_CHANGE_FAILURE;
-        if (!gst_task_start(priv->task.get()))
+        if (!priv->interleave) {
+            gst_element_post_message(element, gst_missing_element_message_new(element, "audiointerleave"));
+            GST_ELEMENT_ERROR(src, CORE, MISSING_PLUGIN, (nullptr), ("no audiointerleave"));
             return GST_STATE_CHANGE_FAILURE;
+        }
+        priv->numberOfSamples = 0;
         break;
-    }
     default:
         break;
     }
@@ -448,6 +461,17 @@ static GstStateChangeReturn webKitWebAudioSrcChangeState(GstElement* element, Gs
     }
 
     switch (transition) {
+    case GST_STATE_CHANGE_READY_TO_PAUSED: {
+        priv->pool = adoptGRef(gst_buffer_pool_new());
+        GstStructure* config = gst_buffer_pool_get_config(priv->pool.get());
+        gst_buffer_pool_config_set_params(config, nullptr, priv->bufferSize, 0, 0);
+        gst_buffer_pool_set_config(priv->pool.get(), config);
+        if (!gst_buffer_pool_set_active(priv->pool.get(), TRUE))
+            returnValue = GST_STATE_CHANGE_FAILURE;
+        else if (!gst_task_start(priv->task.get()))
+            returnValue = GST_STATE_CHANGE_FAILURE;
+        break;
+    }
     case GST_STATE_CHANGE_PAUSED_TO_READY:
         {
             Locker locker { priv->dispatchLock };
